ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +45,255 @@ let tdb_tool = !Xapi_globs.tdb_tool
+        return (l |> List.to_seq |> Map.of_seq)
+
+      let parse_kvp_map (x : string) : (string Map.t, string) result =
+        parse_string ~consume:All kvp_map x
+    end in
+    let ldap fmt = fmt |> Printf.kprintf @@ Printf.sprintf "ldap %s" in
+    let* kvps = P.parse_kvp_map stdout <!> ldap "parsing failed '%s'" in
+    let get_string k =
+      match Map.find_opt k kvps with
+      | None ->
+          Error (ldap "missing key '%s'" k)
+      | Some x ->
+          Ok x
+    in
+    let get_int of_string k =
+      let* str = get_string k in
+      try Ok (of_string str)
+      with _ -> Error (ldap "invalid value for key '%s'" k)
+    in
+    let* upn = get_string "userPrincipalName" in
+    let* user_account_control = get_int Int32.of_string "userAccountControl" in
+    let* account_expires = get_int Int64.of_string "accountExpires" in
+    let account_expired =
+      (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-accountexpires *)
+      let windows_nt_time_to_unix_time x =
+        Int64.sub (Int64.div x 10000000L) 11644473600L
+      in
+      match account_expires with
+      | x when x = 0L || x = Int64.max_int ->
+          false
+      | i ->
+          let expire_unix_time =
+            windows_nt_time_to_unix_time i |> Int64.to_float
+          in
+          expire_unix_time < Unix.time ()
+    in
+    Ok
+      {
+        upn
+      ; account_expired
+        (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-useraccountcontrol
+         * for bit flag docs *)
+      ; account_disabled= Int32.logand user_account_control 2l <> 0l
+      ; account_locked= Int32.logand user_account_control 16l <> 0l
+      ; password_expired= Int32.logand user_account_control 8388608l <> 0l
+      }
+
+  let query_user sid =
+    let query = Printf.sprintf "(&(objectClass=user)(objectSid=%s))" sid in
+    let* stdout =
+      call_ldap query <!> fun _ -> generic_ex "ldap query failed: '%s'" query
+    in
+    parse_user stdout <!> generic_ex "%s"
+end
+
+module Wbinfo = struct
+  let can_resolve_krbtgt () =
+    match call_wbinfo ["-n"; krbtgt] with Ok _ -> true | Error _ -> false
+
+  let sid_of_name name =
+    (* example:
+     *
+     * $ wbinfo -n user@domain.net
+       S-1-2-34-... SID_USER (1)
+     * $ wbinfo -n DOMAIN\user
+       # similar output *)
+    let args = ["-n"; name] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    match String.split_on_char ' ' stdout with
+    | sid :: _ ->
+        Ok (String.trim sid)
+    | [] ->
+        Error
+          (generic_ex "unable to find SID in output of 'wbinfo %s'"
+             (String.concat " " args))
+
+  let name_of_sid =
+    (* example:
+     * $ wbinfo -s S-1-5-21-3143668282-2591278241-912959342-502
+       CONNAPP\krbtgt 1 *)
+    (* we need to drop everything after the last space *)
+    let regex = Re.Perl.(compile (re {|^([^\s].*)\ \d+$|})) in
```

Comment:
per my test, you need to match the ending new line at the end,  something like `let regex = Re.Perl.(compile (re {|^([^\s].*)\ \d+\s$|}))`   `\s` is used to match the ending new line, , tried `\n` for newline, but not work.

Reply:
I couldn't get your regex to work - but both the following work just fine for me:

```bash
# don't allow trailing whitespace
echo "$(wbinfo -s S-1-5-21-1076320343-3137401123-2908695819-456812)" | sed -E -e 's/([^s].*)\ [0-9]+/\1/g'

# do allow trailing whitespace
echo "$(wbinfo -s S-1-5-21-1076320343-3137401123-2908695819-456812)" | sed -E -e 's/([^s].*)\ [0-9]+\s*/\1/g'
```

Reply:
(i've updated the code to allow trailing whitespace)